### PR TITLE
add support for nonnull jsdoc

### LINKED ISSUE #23217: Definite assignment assertions in JSDoc
```js
class ElementWrapper {
  constructor() {
    this.init();
  }

  init() {
    this.element = document.createElement('div');
  }

  getElementStyle() {
    this.element.style
  }
}
```

We need a way to convince TypeScript in `strictNullChecks` that `this.element` is initialized in `getElementStyle`.

### LINKED ISSUE #23405: Support a @nonnullable JSDoc comment
This could be used in place of the non-null assertion operator, and solve #23403.

```js
while (queue.length) {
    (/** @nonnull */ queue.pop())();
}
```

Related is #23217.

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #23405 
Fixes #23217

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -483,6 +483,7 @@ namespace ts {
         JSDocTemplateTag,
         JSDocTypedefTag,
         JSDocPropertyTag,
+        JSDocNonNullTag,
```
Comment (ajafff) [inline src/compiler/types.ts]: `LastJSDocNode` and `LastJSDocTagNode` need to be adjusted

File: src/compiler/utilitiesPublic.ts [hunk 2]
```diff
@@ -1622,6 +1626,10 @@ namespace ts {
         return node.kind === SyntaxKind.JSDocTypeTag;
     }
 
+    export function jsJSDocNonNullTag(node: Node): node is JSDocNonNullTag {
```
Comment (ajafff) [inline src/compiler/utilitiesPublic.ts]: typo: `js` -> `is`

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -32376,7 +32375,7 @@ namespace ts {
                 if (getModifierFlags(member) & ModifierFlags.Ambient) {
                     continue;
                 }
-                if (isInstancePropertyWithoutInitializer(member)) {
+                if (isInstancePropertyWithoutInitializer(member) && !isNonNullPropertyWithJSDoc(member)) {
```
Comment (ajafff) [inline src/compiler/checker.ts]: is this the equivalent of a "definite assignment assertion" on properties?
If so, does this need to be expanded to variables as well?

**Comment by DanielRosenwasser** [comment]
The way this works is kind of unfortunate because I don't think you can mix/match these with optional chaining the way we're changing in TypeScript 3.8. (@rbuckton)

So you can't easily convert to strictNullChecks by switching

```js
a?.b.c
```

to

```js
(/** @nonnull */ (a?.b)).c
```

**Comment by rbuckton** [comment]
We could instead look for it in the trailing comments of a node:

```js
a?.b/** @nonnull */.c
```

**Comment by ajafff** [comment]
@DanielRosenwasser why would you ever want to use a nonnull assertion in an optional chain? There's another issue discussing whether to make that an error.
I just want to know if this is a valid use case that needs to be considered here

**Comment by DanielRosenwasser** [comment]
@ajaff if you have `a?.b.c`, `b` might be undefined and you might not want to propagate that value out with another optional chain because you're asserting that it *should* be there - it's an error condition if it's not. So you might want to write `a?.b!.c`.

**Comment by ajafff** [comment]
@DanielRosenwasser so you mean "Optional Chaining with Non-Null Assertions" that was discussed in #36138? Now I understand what you mean.

---

> We could instead look for it in the trailing comments of a node

AFAICT that would require parsing JSDoc for **every** expression. In addition this requires the JSDoc parser to use the trailing comments (and store them in a separate place).
And on top of that, the checker needs to look for JSDoc on every expression when computing contextual types or actual types.

**Pull request closed without merging**